In [1]:
import pandas as pd
import numpy as np
# Imputación de nulos usando métodos avanzados estadísticos
# -----------------------------------------------------------------------
from sklearn.impute import SimpleImputer
from sklearn.experimental import enable_iterative_imputer
from sklearn.impute import IterativeImputer
from sklearn.impute import KNNImputer

# Librerías de visualización
# -----------------------------------------------------------------------
import seaborn as sns
import matplotlib.pyplot as plt
# Configuración
# -----------------------------------------------------------------------


1 - Importa Pandas


- - En el inicio del jupiter se han importado todas las librería que pueden ser útiles para analizar, limpiar y modificar los datos

**2-  Exploración del conjunto de datos:**

2.1 - Carga el fichero, al cargarlo te aparecerá una columna llamada "Unnamed: 0". Carga el dataset sin que aparezca esta columna. Además realiza un análisis exploratorio que incluya:


In [2]:
df = pd.read_csv("Spotify_Youtube.csv", index_col=0)

2.2 - Cuando leas el fichero, veras que no podemos ver todas las columnas, utiliza el comando correcto para poder visualizarlas todas cuando hacemos un head


In [3]:
pd.set_option('display.max_columns', None) # para poder visualizar todas las columnas de los DataFrames

2.3 - Primeras 5 filas del dataframe


In [4]:
df.head(5)

Artist                                        Url_spotify  \
0  Gorillaz  https://open.spotify.com/artist/3AA28KZvwAUcZu...   
1  Gorillaz  https://open.spotify.com/artist/3AA28KZvwAUcZu...   
2  Gorillaz  https://open.spotify.com/artist/3AA28KZvwAUcZu...   
3  Gorillaz  https://open.spotify.com/artist/3AA28KZvwAUcZu...   
4  Gorillaz  https://open.spotify.com/artist/3AA28KZvwAUcZu...   

                                           Track  \
0                                 Feel Good Inc.   
1                                Rhinestone Eyes   
2  New Gold (feat. Tame Impala and Bootie Brown)   
3                             On Melancholy Hill   
4                                 Clint Eastwood   

                                           Album Album_type  \
0                                     Demon Days      album   
1                                  Plastic Beach      album   
2  New Gold (feat. Tame Impala and Bootie Brown)     single   
3                                  Plastic Beach      album   
4                                       Gorillaz      album   

                                    Uri Danceability Energy   Key Loudness  \
0  spotify:track:0d28khcov6AiegSCpG5TuT        0,818  0,705   6,0   -6,679   
1  spotify:track:1foMv2HQwfQ2vntFf9HFeG        0,676  0,703   8,0   -5,815   
2  spotify:track:64dLd6rVqDLtkXFYrEUHIU        0,695  0,923   1,0    -3,93   
3  spotify:track:0q6LuUqGLUiCPP1cbdwFs3        0,689  0,739   2,0    -5,81   
4  spotify:track:7yMiX7n9SBvadzox8T5jzT        0,663  0,694  10,0   -8,627   

  Speechiness Acousticness Instrumentalness  Liveness  Valence    Tempo  \
0       0,177      0,00836          0,00233    0.6130    0.772  138.559   
1      0,0302       0,0869         0,000687    0.0463    0.852   92.761   
2      0,0522       0,0425           0,0469    0.1160    0.551  108.014   
3       0,026     1,51e-05            0,509    0.0640    0.578  120.423   
4       0,171       0,0253              0,0    0.0698    0.525  167.953   

   Duration_ms                                  Url_youtube  \
0     222640.0  https://www.youtube.com/watch?v=HyHNuVaZJ-k   
1     200173.0  https://www.youtube.com/watch?v=yYDmaexVHic   
2     215150.0  https://www.youtube.com/watch?v=qJa-VFwPpYA   
3     233867.0  https://www.youtube.com/watch?v=04mfKJWDSzI   
4     340920.0  https://www.youtube.com/watch?v=1V_xRb0x9aw   

                                               Title   Channel        Views  \
0         Gorillaz - Feel Good Inc. (Official Video)  Gorillaz  693555221.0   
1  Gorillaz - Rhinestone Eyes [Storyboard Film] (...  Gorillaz   72011645.0   
2  Gorillaz - New Gold ft. Tame Impala & Bootie B...  Gorillaz    8435055.0   
3     Gorillaz - On Melancholy Hill (Official Video)  Gorillaz  211754952.0   
4         Gorillaz - Clint Eastwood (Official Video)  Gorillaz  618480958.0   

       Likes  Comments                                        Description  \
0  6220896.0  169907.0  Official HD Video for Gorillaz' fantastic trac...   
1  1079128.0   31003.0  The official video for Gorillaz - Rhinestone E...   
2   282142.0    7399.0  Gorillaz - New Gold ft. Tame Impala & Bootie B...   
3  1788577.0   55229.0  Follow Gorillaz online:\nhttp://gorillaz.com \...   
4  6197318.0  155930.0  The official music video for Gorillaz - Clint ...   

  Licensed Official_video        Stream  
0     True           True  1.040235e+09  
1     True           True  3.100837e+08  
2     True           True  6.306347e+07  
3     True           True  4.346636e+08  
4     True           True  6.172597e+08

2.4 Últimas 5 filas del dataframe


In [5]:
df.tail(5)

Artist                                        Url_spotify  \
20713  SICK LEGEND  https://open.spotify.com/artist/3EYY5FwDkHEYLw...   
20714  SICK LEGEND  https://open.spotify.com/artist/3EYY5FwDkHEYLw...   
20715  SICK LEGEND  https://open.spotify.com/artist/3EYY5FwDkHEYLw...   
20716  SICK LEGEND  https://open.spotify.com/artist/3EYY5FwDkHEYLw...   
20717  SICK LEGEND  https://open.spotify.com/artist/3EYY5FwDkHEYLw...   

                                Track                           Album  \
20713            JUST DANCE HARDSTYLE            JUST DANCE HARDSTYLE   
20714  SET FIRE TO THE RAIN HARDSTYLE  SET FIRE TO THE RAIN HARDSTYLE   
20715       OUTSIDE HARDSTYLE SPED UP       OUTSIDE HARDSTYLE SPED UP   
20716             ONLY GIRL HARDSTYLE             ONLY GIRL HARDSTYLE   
20717              MISS YOU HARDSTYLE              MISS YOU HARDSTYLE   

      Album_type                                   Uri Danceability Energy  \
20713     single  spotify:track:0RtcKQGyI4hr8FgFH1TuYG        0,582  0,926   
20714     single  spotify:track:3rHvPA8lUnPBkaLyPOc0VV        0,531  0,936   
20715     single  spotify:track:4jk00YxPtPbhvHJE9N4ddv        0,443   0,83   
20716     single  spotify:track:5EyErbpsugWliX006eTDex        0,417  0,767   
20717     single  spotify:track:6lOn0jz1QpjcWeXo1oMm0k        0,498  0,938   

       Key Loudness Speechiness Acousticness Instrumentalness  Liveness  \
20713  5,0   -6,344      0,0328        0,448              0,0    0.0839   
20714  4,0   -1,786       0,137        0,028              0,0    0.0923   
20715  4,0   -4,679      0,0647       0,0243              0,0    0.1540   
20716  9,0   -4,004       0,419        0,356           0,0184    0.1080   
20717  6,0   -4,543       0,107      0,00277            0,911    0.1360   

       Valence    Tempo  Duration_ms  \
20713   0.6580   90.002      94667.0   
20714   0.6570  174.869     150857.0   
20715   0.4190  168.388     136842.0   
20716   0.5390  155.378     108387.0   
20717   0.0787  160.067     181500.0   

                                       Url_youtube  \
20713  https://www.youtube.com/watch?v=5SHmKFKlNqI   
20714  https://www.youtube.com/watch?v=ocTH6KxllDQ   
20715  https://www.youtube.com/watch?v=5wFhE-HY0hg   
20716  https://www.youtube.com/watch?v=VMFLbFRNCn0   
20717  https://www.youtube.com/watch?v=zau0dckCFi0   

                                Title              Channel     Views   Likes  \
20713            JUST DANCE HARDSTYLE  SICK LEGEND - Topic   71678.0  1113.0   
20714  SET FIRE TO THE RAIN HARDSTYLE  SICK LEGEND - Topic  164741.0  2019.0   
20715       OUTSIDE HARDSTYLE SPED UP  SICK LEGEND - Topic   35646.0   329.0   
20716             ONLY GIRL HARDSTYLE  SICK LEGEND - Topic    6533.0    88.0   
20717              MISS YOU HARDSTYLE  SICK LEGEND - Topic  158697.0  2484.0   

       Comments                                        Description Licensed  \
20713       0.0  Provided to YouTube by Routenote\n\nJUST DANCE...     True   
20714       0.0  Provided to YouTube by Routenote\n\nSET FIRE T...     True   
20715       0.0  Provided to YouTube by Routenote\n\nOUTSIDE HA...     True   
20716       0.0  Provided to YouTube by Routenote\n\nONLY GIRL ...     True   
20717       0.0  Provided to YouTube by Routenote\n\nMISS YOU H...     True   

      Official_video      Stream  
20713           True   9227144.0  
20714           True  10898176.0  
20715           True   6226110.0  
20716           True   6873961.0  
20717           True   5695584.0

2.5 - 10 filas aleatorias del dataframe


In [6]:
df.sample(10)

Artist  \
9859              Tulsi Kumar   
2158              Wyclef Jean   
17856  Los Gemelos De Sinaloa   
910                    T-Pain   
8441           Oliver Heldens   
8072                    TWICE   
17059         Miggy Dela Rosa   
12173              Thiaguinho   
3126                     E-40   
2836           Roberto Carlos   

                                             Url_spotify  \
9859   https://open.spotify.com/artist/0T1CMVkqffHlqE...   
2158   https://open.spotify.com/artist/7aBzpmFXB4WWpP...   
17856  https://open.spotify.com/artist/1Zkxm1dM3HI3Qk...   
910    https://open.spotify.com/artist/3aQeKQSyrW4qWr...   
8441   https://open.spotify.com/artist/5nki7yRhxgM509...   
8072   https://open.spotify.com/artist/7n2Ycct7Beij7D...   
17059  https://open.spotify.com/artist/45ruzGUmIr8WLj...   
12173  https://open.spotify.com/artist/1vppDmG3i5sXf3...   
3126   https://open.spotify.com/artist/3crnzLy8R4lVwa...   
2836   https://open.spotify.com/artist/7fAKtXSdNInWAI...   

                                                  Track  \
9859                                            Shut Up   
2158                                          Fu-Gee-La   
17856                                      No Le Aflojo   
910                                          I'm Sprung   
8441                                            Mercury   
8072                                    I CAN'T STOP ME   
17059                                   Ferrari - Remix   
12173                             Ponto Fraco - Ao Vivo   
3126             U and Dat (feat. T. Pain & Kandi Girl)   
2836   El Gato Que Está Triste y Azul - Un Gato Nel Blu   

                                      Album Album_type  \
9859                                Shut Up     single   
2158           The Score (Expanded Edition)      album   
17856                          No Le Aflojo     single   
910    Rappa Ternt Sanga (Expanded Edition)      album   
8441                                Mercury     single   
8072                         Eyes Wide Open      album   
17059                       Ferrari (Remix)     single   
12173            Só Vem! (ao Vivo) - Deluxe      album   
3126                  My Ghetto Report Card      album   
2836                   30 Grandes Canciones      album   

                                        Uri Danceability Energy   Key  \
9859   spotify:track:4OvAutMwVCBfGspBieI9i0        0,821  0,761  10,0   
2158   spotify:track:1rJM37Dd4OcSJvzI84yCcP        0,681  0,505   1,0   
17856  spotify:track:7ifMxHITc3NIipJO4R5NGb        0,698  0,687   5,0   
910    spotify:track:0DLOyyQvwPTSDKuhpzMMwA        0,728  0,327   0,0   
8441   spotify:track:25dFcH5EsCJLMSmFDDJLT3         0,78  0,986   7,0   
8072   spotify:track:37ZtpRBkHcaq6hHy0X98zn        0,657   0,88  10,0   
17059  spotify:track:3XucsgiwXb8KPn9Csf9Zmu        0,818  0,728   1,0   
12173  spotify:track:0KkZFQSvZrTL1zU3cX8Z7e        0,605  0,827   6,0   
3126   spotify:track:7g7M96UGiA7OR79H86gV4m        0,858  0,484   6,0   
2836   spotify:track:3SrnZHhXpgtvHerbxEDOWW        0,447  0,457   4,0   

      Loudness Speechiness Acousticness Instrumentalness  Liveness  Valence  \
9859    -4,815      0,0619       0,0445         4,33e-05    0.0661    0.780   
2158   -11,709       0,138        0,137              0,0    0.2340    0.399   
17856   -4,115      0,0501        0,195              0,0    0.0829    0.773   
910    -11,596      0,0999       0,0948              0,0    0.1030    0.195   
8441    -3,446      0,0594       0,0214            0,851    0.1200    0.397   
8072    -3,672       0,102        0,121              0,0    0.4630    0.769   
17059   -5,705      0,0724       0,0985              0,0    0.0917    0.597   
12173   -6,635      0,0503        0,725              0,0    0.9520    0.785   
3126    -7,448       0,277       0,0241          1,8e-06    0.1230    0.329   
2836   -10,969      0,0331        0,675              0,0    0.3390    0.590   

         Tempo  Duration_ms        

2.6 - ¿Cuántas filas y columnas tenemos en el dataframe


In [7]:
df.shape
# Hay 20718 filas y 27 columnas

(20718, 27)

2.7 - ¿Cuáles son los tipos de los datos de cada columna del dataframe


In [8]:
df.info()
# Vemos que hay 8 columnas tipo float, y 19 tipo object
# Esto nos indica que probablemente tengamos que transformar algunas columnas de obj a float

<class 'pandas.core.frame.DataFrame'>
Index: 20718 entries, 0 to 20717
Data columns (total 27 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   Artist            20718 non-null  object 
 1   Url_spotify       20718 non-null  object 
 2   Track             20718 non-null  object 
 3   Album             20718 non-null  object 
 4   Album_type        20718 non-null  object 
 5   Uri               20718 non-null  object 
 6   Danceability      20716 non-null  object 
 7   Energy            20716 non-null  object 
 8   Key               20716 non-null  object 
 9   Loudness          20716 non-null  object 
 10  Speechiness       20716 non-null  object 
 11  Acousticness      20716 non-null  object 
 12  Instrumentalness  20716 non-null  object 
 13  Liveness          20716 non-null  float64
 14  Valence           20716 non-null  float64
 15  Tempo             20716 non-null  float64
 16  Duration_ms       20716 non-null  float64
 17

2.8 - ¿Cuántos valores nulos tenemos por columna?


In [9]:
# 1 miramos el porcentaje de nulos:
df_nulos =pd.DataFrame(df.isnull().sum() / df.shape[0]) * 100
df_nulos

0
Artist            0.000000
Url_spotify       0.000000
Track             0.000000
Album             0.000000
Album_type        0.000000
Uri               0.000000
Danceability      0.009653
Energy            0.009653
Key               0.009653
Loudness          0.009653
Speechiness       0.009653
Acousticness      0.009653
Instrumentalness  0.009653
Liveness          0.009653
Valence           0.009653
Tempo             0.009653
Duration_ms       0.009653
Url_youtube       2.268559
Title             2.268559
Channel           2.268559
Views             2.268559
Likes             2.611256
Comments          2.746404
Description       4.228207
Licensed          2.268559
Official_video    2.268559
Stream            2.780191

In [10]:
#2: filtramos las columnas que  tengan nulos:
df_nulos[df_nulos[0] > 0]

0
Danceability      0.009653
Energy            0.009653
Key               0.009653
Loudness          0.009653
Speechiness       0.009653
Acousticness      0.009653
Instrumentalness  0.009653
Liveness          0.009653
Valence           0.009653
Tempo             0.009653
Duration_ms       0.009653
Url_youtube       2.268559
Title             2.268559
Channel           2.268559
Views             2.268559
Likes             2.611256
Comments          2.746404
Description       4.228207
Licensed          2.268559
Official_video    2.268559
Stream            2.780191

2.9 - ¿Tenemos filas duplicadas en el dataframe ?


In [11]:
df.duplicated().sum()

# No hay filas duplicadas

0

2.10 -  Muestra los principales estadísticos para las columnas numéricas del dataframe


In [12]:
# principales estadisticos de las columnas numéricas

display(df.describe(exclude = "O").T)

count          mean           std         min           25%  \
Liveness     20716.0  1.935210e-01  1.685309e-01      0.0145  9.410000e-02   
Valence      20716.0  5.298533e-01  2.454408e-01      0.0000  3.390000e-01   
Tempo        20716.0  1.206383e+02  2.957902e+01      0.0000  9.700200e+01   
Duration_ms  20716.0  2.247176e+05  1.247905e+05  30985.0000  1.800095e+05   
Views        20248.0  9.393782e+07  2.746443e+08      0.0000  1.826002e+06   
Likes        20177.0  6.633411e+05  1.789324e+06      0.0000  2.158100e+04   
Comments     20149.0  2.751899e+04  1.932347e+05      0.0000  5.090000e+02   
Stream       20142.0  1.359422e+08  2.441321e+08   6574.0000  1.767486e+07   

                      50%           75%           max  
Liveness     1.250000e-01  2.370000e-01  1.000000e+00  
Valence      5.370000e-01  7.262500e-01  9.930000e-01  
Tempo        1.199650e+02  1.399350e+02  2.433720e+02  
Duration_ms  2.132845e+05  2.524430e+05  4.676058e+06  
Views        1.450110e+07  7.039975e+07  8.079649e+09  
Likes        1.244810e+05  5.221480e+05  5.078865e+07  
Comments     3.277000e+03  1.436000e+04  1.608314e+07  
Stream       4.968298e+07  1.383581e+08  3.386520e+09

2.11 -  Muestra los principales estadísticos para las columnas categóricas del dataframe

In [13]:
# principales estadisticos de las columnas numéricas

display(df.describe(include = "O", exclude='int').T)

count unique  \
Artist            20718   2079   
Url_spotify       20718   2079   
Track             20718  17841   
Album             20718  11937   
Album_type        20718      3   
Uri               20718  18862   
Danceability      20716    898   
Energy            20716   1268   
Key               20716     12   
Loudness          20716   9417   
Speechiness       20716   1303   
Acousticness      20716   3138   
Instrumentalness  20716   4012   
Url_youtube       20248  18154   
Title             20248  18146   
Channel           20248   6714   
Description       19842  17395   
Licensed          20248      2   
Official_video    20248      2   

                                                                top   freq  
Artist                                                     Gorillaz     10  
Url_spotify       https://open.spotify.com/artist/3AA28KZvwAUcZu...     10  
Track                     El Ultimo Adiós - Varios Artistas Version     24  
Album                                                 Greatest Hits     30  
Album_type                                                    album  14926  
Uri                            spotify:track:0tzixmHNQfE6S6SirSToxW     24  
Danceability                                                  0,687     78  
Energy                                                        0,572     60  
Key                                                             0,0   2305  
Loudness                                                     -7,818     25  
Speechiness                                                  0,0324     72  
Acousticness                                                  0,114     50  
Instrumentalness                                                0,0   9391  
Url_youtube             https://www.youtube.com/watch?v=GqOvxbFLwjY     18  
Title             Color Esperanza 2020 - Various Artists (Offici...     18  
Channel                                                    T-Series    238  
Description                                                            114  
Licensed                                                       True  14140  
Official_video                                                 True  15723

In [14]:
# seleccionamos variables categoricas para ver sus valores unicos:
df_categoricos = df.select_dtypes(include = "O")
# con .T cambiamos las filas por las colmnas (tansponer)
df_categoricos.head()

for col in df_categoricos.columns:
    print("Los valores estadísticos para la columna: ", col.upper())
    display(df[col].describe().reset_index().head())

Los valores estadísticos para la columna:  ARTIST


index    Artist
0   count     20718
1  unique      2079
2     top  Gorillaz
3    freq        10

Los valores estadísticos para la columna:  URL_SPOTIFY


index                                        Url_spotify
0   count                                              20718
1  unique                                               2079
2     top  https://open.spotify.com/artist/3AA28KZvwAUcZu...
3    freq                                                 10

Los valores estadísticos para la columna:  TRACK


index                                      Track
0   count                                      20718
1  unique                                      17841
2     top  El Ultimo Adiós - Varios Artistas Version
3    freq                                         24

Los valores estadísticos para la columna:  ALBUM


index          Album
0   count          20718
1  unique          11937
2     top  Greatest Hits
3    freq             30

Los valores estadísticos para la columna:  ALBUM_TYPE


index Album_type
0   count      20718
1  unique          3
2     top      album
3    freq      14926

Los valores estadísticos para la columna:  URI


index                                   Uri
0   count                                 20718
1  unique                                 18862
2     top  spotify:track:0tzixmHNQfE6S6SirSToxW
3    freq                                    24

Los valores estadísticos para la columna:  DANCEABILITY


index Danceability
0   count        20716
1  unique          898
2     top        0,687
3    freq           78

Los valores estadísticos para la columna:  ENERGY


index Energy
0   count  20716
1  unique   1268
2     top  0,572
3    freq     60

Los valores estadísticos para la columna:  KEY


index    Key
0   count  20716
1  unique     12
2     top    0,0
3    freq   2305

Los valores estadísticos para la columna:  LOUDNESS


index Loudness
0   count    20716
1  unique     9417
2     top   -7,818
3    freq       25

Los valores estadísticos para la columna:  SPEECHINESS


index Speechiness
0   count       20716
1  unique        1303
2     top      0,0324
3    freq          72

Los valores estadísticos para la columna:  ACOUSTICNESS


index Acousticness
0   count        20716
1  unique         3138
2     top        0,114
3    freq           50

Los valores estadísticos para la columna:  INSTRUMENTALNESS


index Instrumentalness
0   count            20716
1  unique             4012
2     top              0,0
3    freq             9391

Los valores estadísticos para la columna:  URL_YOUTUBE


index                                  Url_youtube
0   count                                        20248
1  unique                                        18154
2     top  https://www.youtube.com/watch?v=GqOvxbFLwjY
3    freq                                           18

Los valores estadísticos para la columna:  TITLE


index                                              Title
0   count                                              20248
1  unique                                              18146
2     top  Color Esperanza 2020 - Various Artists (Offici...
3    freq                                                 18

Los valores estadísticos para la columna:  CHANNEL


index   Channel
0   count     20248
1  unique      6714
2     top  T-Series
3    freq       238

Los valores estadísticos para la columna:  DESCRIPTION


index Description
0   count       19842
1  unique       17395
2     top            
3    freq         114

Los valores estadísticos para la columna:  LICENSED


index Licensed
0   count    20248
1  unique        2
2     top     True
3    freq    14140

Los valores estadísticos para la columna:  OFFICIAL_VIDEO


index Official_video
0   count          20248
1  unique              2
2     top           True
3    freq          15723

____

**3. Preparación de los datos:**

3.1 - Los nombres de las columnas empiezan con mayúsculas, pon todos los nombres de las columnas en minúsculas.

In [15]:
df.columns

Index(['Artist', 'Url_spotify', 'Track', 'Album', 'Album_type', 'Uri',
       'Danceability', 'Energy', 'Key', 'Loudness', 'Speechiness',
       'Acousticness', 'Instrumentalness', 'Liveness', 'Valence', 'Tempo',
       'Duration_ms', 'Url_youtube', 'Title', 'Channel', 'Views', 'Likes',
       'Comments', 'Description', 'Licensed', 'Official_video', 'Stream'],
      dtype='object')

In [16]:
df.columns = df.columns.str.lower()

In [17]:
df.head(5)

artist                                        url_spotify  \
0  Gorillaz  https://open.spotify.com/artist/3AA28KZvwAUcZu...   
1  Gorillaz  https://open.spotify.com/artist/3AA28KZvwAUcZu...   
2  Gorillaz  https://open.spotify.com/artist/3AA28KZvwAUcZu...   
3  Gorillaz  https://open.spotify.com/artist/3AA28KZvwAUcZu...   
4  Gorillaz  https://open.spotify.com/artist/3AA28KZvwAUcZu...   

                                           track  \
0                                 Feel Good Inc.   
1                                Rhinestone Eyes   
2  New Gold (feat. Tame Impala and Bootie Brown)   
3                             On Melancholy Hill   
4                                 Clint Eastwood   

                                           album album_type  \
0                                     Demon Days      album   
1                                  Plastic Beach      album   
2  New Gold (feat. Tame Impala and Bootie Brown)     single   
3                                  Plastic Beach      album   
4                                       Gorillaz      album   

                                    uri danceability energy   key loudness  \
0  spotify:track:0d28khcov6AiegSCpG5TuT        0,818  0,705   6,0   -6,679   
1  spotify:track:1foMv2HQwfQ2vntFf9HFeG        0,676  0,703   8,0   -5,815   
2  spotify:track:64dLd6rVqDLtkXFYrEUHIU        0,695  0,923   1,0    -3,93   
3  spotify:track:0q6LuUqGLUiCPP1cbdwFs3        0,689  0,739   2,0    -5,81   
4  spotify:track:7yMiX7n9SBvadzox8T5jzT        0,663  0,694  10,0   -8,627   

  speechiness acousticness instrumentalness  liveness  valence    tempo  \
0       0,177      0,00836          0,00233    0.6130    0.772  138.559   
1      0,0302       0,0869         0,000687    0.0463    0.852   92.761   
2      0,0522       0,0425           0,0469    0.1160    0.551  108.014   
3       0,026     1,51e-05            0,509    0.0640    0.578  120.423   
4       0,171       0,0253              0,0    0.0698    0.525  167.953   

   duration_ms                                  url_youtube  \
0     222640.0  https://www.youtube.com/watch?v=HyHNuVaZJ-k   
1     200173.0  https://www.youtube.com/watch?v=yYDmaexVHic   
2     215150.0  https://www.youtube.com/watch?v=qJa-VFwPpYA   
3     233867.0  https://www.youtube.com/watch?v=04mfKJWDSzI   
4     340920.0  https://www.youtube.com/watch?v=1V_xRb0x9aw   

                                               title   channel        views  \
0         Gorillaz - Feel Good Inc. (Official Video)  Gorillaz  693555221.0   
1  Gorillaz - Rhinestone Eyes [Storyboard Film] (...  Gorillaz   72011645.0   
2  Gorillaz - New Gold ft. Tame Impala & Bootie B...  Gorillaz    8435055.0   
3     Gorillaz - On Melancholy Hill (Official Video)  Gorillaz  211754952.0   
4         Gorillaz - Clint Eastwood (Official Video)  Gorillaz  618480958.0   

       likes  comments                                        description  \
0  6220896.0  169907.0  Official HD Video for Gorillaz' fantastic trac...   
1  1079128.0   31003.0  The official video for Gorillaz - Rhinestone E...   
2   282142.0    7399.0  Gorillaz - New Gold ft. Tame Impala & Bootie B...   
3  1788577.0   55229.0  Follow Gorillaz online:\nhttp://gorillaz.com \...   
4  6197318.0  155930.0  The official music video for Gorillaz - Clint ...   

  licensed official_video        stream  
0     True           True  1.040235e+09  
1     True           True  3.100837e+08  
2     True           True  6.306347e+07  
3     True           True  4.346636e+08  
4     True           True  6.172597e+08

3.2 - Haciendo el análisis exploratorio te deberías haber dado cuenta de que algunas de las variables (danceability, energy, key, loudness, speechiness, acousticness, instrumentalness) no son del tipo que deberían. Esto es debido a que los decimales están establecidos como comas y no con puntos. Crea una función que nos permita cambiar esas comas por puntos para que los datos tengan el tipo correcto.

In [18]:
df["danceability"]

0        0,818
1        0,676
2        0,695
3        0,689
4        0,663
         ...  
20713    0,582
20714    0,531
20715    0,443
20716    0,417
20717    0,498
Name: danceability, Length: 20718, dtype: object

In [19]:
def cambiar_comas(cadena):
    try:
        # Reemplazar las comas por puntos en la cadena
        return float(cadena.replace(",", "."))
    
    except:
        # Si ocurre algún error (por ejemplo, si el argumento no es una cadena),
        # devolver np.nan (valor Not a Number) para indicar un valor inválido o no disponible.
        return np.nan
    

In [20]:
columnas_a_float = ["danceability","energy", "key", "loudness", "speechiness", "acousticness", "instrumentalness" ]

# iteramos por la lista de columnas y a cada una de ellas le aplicamos la función que hemos creado con un '.apply()'
for col in columnas_a_float:
    df[col] = df[col].apply(cambiar_comas)

3.3 - Después de haber hecho los cambios, chequea los tipos de datos. ¿Son ya del tipo correcto? En caso de que no, crea otra función o modifica la anterior para que sean de tipo float.

In [21]:
# comprobamos los tipos de datos -> Todos han pasado a float
df[columnas_a_float].info()

<class 'pandas.core.frame.DataFrame'>
Index: 20718 entries, 0 to 20717
Data columns (total 7 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   danceability      20716 non-null  float64
 1   energy            20716 non-null  float64
 2   key               20716 non-null  float64
 3   loudness          20716 non-null  float64
 4   speechiness       20716 non-null  float64
 5   acousticness      20716 non-null  float64
 6   instrumentalness  20716 non-null  float64
dtypes: float64(7)
memory usage: 1.3 MB


3.4 - Hay algunas canciones cuyo título está en mayúsulas. Crea una función para que todos los títulos estén en minúscula. Haz lo mismo para las columnas de "title" y "album".

In [22]:
df[['title', 'album']]

title  \
0             Gorillaz - Feel Good Inc. (Official Video)   
1      Gorillaz - Rhinestone Eyes [Storyboard Film] (...   
2      Gorillaz - New Gold ft. Tame Impala & Bootie B...   
3         Gorillaz - On Melancholy Hill (Official Video)   
4             Gorillaz - Clint Eastwood (Official Video)   
...                                                  ...   
20713                               JUST DANCE HARDSTYLE   
20714                     SET FIRE TO THE RAIN HARDSTYLE   
20715                          OUTSIDE HARDSTYLE SPED UP   
20716                                ONLY GIRL HARDSTYLE   
20717                                 MISS YOU HARDSTYLE   

                                               album  
0                                         Demon Days  
1                                      Plastic Beach  
2      New Gold (feat. Tame Impala and Bootie Brown)  
3                                      Plastic Beach  
4                                           Gorillaz  
...                                              ...  
20713                           JUST DANCE HARDSTYLE  
20714                 SET FIRE TO THE RAIN HARDSTYLE  
20715                      OUTSIDE HARDSTYLE SPED UP  
20716                            ONLY GIRL HARDSTYLE  
20717                             MISS YOU HARDSTYLE  

[20718 rows x 2 columns]

In [23]:
col_datos_a_min = ['title', 'album']

df[col_datos_a_min] = df[col_datos_a_min].apply(lambda texto: texto.str.lower())



In [24]:
df[['title', 'album']]

title  \
0             gorillaz - feel good inc. (official video)   
1      gorillaz - rhinestone eyes [storyboard film] (...   
2      gorillaz - new gold ft. tame impala & bootie b...   
3         gorillaz - on melancholy hill (official video)   
4             gorillaz - clint eastwood (official video)   
...                                                  ...   
20713                               just dance hardstyle   
20714                     set fire to the rain hardstyle   
20715                          outside hardstyle sped up   
20716                                only girl hardstyle   
20717                                 miss you hardstyle   

                                               album  
0                                         demon days  
1                                      plastic beach  
2      new gold (feat. tame impala and bootie brown)  
3                                      plastic beach  
4                                           gorillaz  
...                                              ...  
20713                           just dance hardstyle  
20714                 set fire to the rain hardstyle  
20715                      outside hardstyle sped up  
20716                            only girl hardstyle  
20717                             miss you hardstyle  

[20718 rows x 2 columns]

3.5 - Algunos de los artistas tienen símbolos raros, en concreto el símbolo $. Crea una función que los reemplaze pos "s".

nota: no me ha salido correctamente

In [25]:
df['artist'].values

array(['Gorillaz', 'Gorillaz', 'Gorillaz', ..., 'SICK LEGEND',
       'SICK LEGEND', 'SICK LEGEND'], dtype=object)

In [26]:
#1: detectamos que artistas tienen el símbolo $ en el nombre

patron = r'\b\w*\$\w*\b'
df_con_dolar = df[df['artist'].str.contains(patron, regex=True, na=False)]
df_con_dolar.sample(5)


artist                                        url_spotify  \
10256    Curren$y  https://open.spotify.com/artist/6X8WdFjrNhXATM...   
15315  A$AP Rocky  https://open.spotify.com/artist/13ubrt8QOOCPlj...   
15646   A$AP Ferg  https://open.spotify.com/artist/5dHt1vcEm9qb8f...   
15640   A$AP Ferg  https://open.spotify.com/artist/5dHt1vcEm9qb8f...   
10265    Curren$y  https://open.spotify.com/artist/6X8WdFjrNhXATM...   

                                      track  \
10256                    3AM in New Orleans   
15315                                   L$D   
15646              New Level (feat. Future)   
15640  Plain Jane REMIX (feat. Nicki Minaj)   
10265                           Radioactive   

                                      album album_type  \
10256               lil black jean jacket 3      album   
15315                     at.long.last.a$ap      album   
15646             always strive and prosper      album   
15640  plain jane remix (feat. nicki minaj)     single   
10265           the drive in theatre part 2      album   

                                        uri  danceability  energy   key  \
10256  spotify:track:18kOgPYJgcMCpurna8P23T         0.649   0.647   6.0   
15315  spotify:track:4S7YHmlWwfwArgd8LfSPud         0.598   0.252   2.0   
15646  spotify:track:25rc1PFgPvtV6GFSloyOI0         0.734   0.703   1.0   
15640  spotify:track:2YzmfPLqUx5CJOaw5ThsBV         0.791   0.839  11.0   
10265  spotify:track:45AIuD8zNP8MzT7KyuoBcA         0.748   0.792  11.0   

       loudness  speechiness  acousticness  instrumentalness  liveness  \
10256    -7.386       0.0913        0.0496          0.000800    0.1040   
15315   -15.086       0.0644        0.1940          0.133000    0.0992   
15646    -1.690       0.1320        0.0344          0.000000    0.1620   
15640    -4.284       0.2660        0.0588          0.000063    0.2590   
10265    -5.180       0.2730        0.0950          0.000000    0.1660   

       valence    tempo  duration_ms  \
10256   0.8440  158.982     133679.0   
15315   0.0352  120.064     238307.0   
15646   0.1520  121.080     264973.0   
15640   0.3830  170.113     202636.0   
10265   0.5400  134.041     188291.0   

                                       url_youtube  \
10256  https://www.youtube.com/watch?v=m85jepqoXR0   
15315  https://www.youtube.com/watch?v=yEG2VTHS9yg   
15646  https://www.youtube.com/watch?v=Srns7NiO278   
15640  https://www.youtube.com/watch?v=XDMGmtpvBjs   
10265  https://www.youtube.com/watch?v=B-zzjJGQa9k   

                                                   title           channel  \
10256  a$ap ant & curren$y - 3am in new orleans (offi...   MARINO INFANTRY   
15315             a$ap rocky - l$d (love x $ex x dreams)   ASAPROCKYUPTOWN   
15646  a$ap ferg - new level (official video) ft. future      asapfergVEVO   
15640  a$ap ferg - plain jane remix (official audio) ...      asapfergVEVO   
10265                                        radioactive  Curren$y - Topic   

             views      likes  comments  \
10256    1209868.0    31623.0     816.0   
15315  179436818.0  1404266.0   45042.0   
15646  104920979.0   749589.0   21900.0   
15640  176071336.0  1679681.0   32747.0   
10265     145096.0     3034.0      16.0   

                                             description licensed  \
10256  Track # 8 featuring Curren$y off A$AP Ant's La...     True   
15315  Directed By Dexter Navy\nCo-Directed By A$AP R...     True   
15646  Official music video for ”New Level” by A$AP F...     True   
15640  "Plain Jane REMIX” by A$AP Ferg feat. Nicki Mi...     True   
10265  Provided to YouTube by EMPIRE Distribution\n\n...     True   

      official_video       stream  
10256           True   11975020.0  
15315           True  452518080.0  
15646           True  174252080.0  
15640           True  248818454.0  
10265           True     463086.0

In [47]:
def reemplazar_simbolo_dolar(nombre):
    try: 
        return nombre.replace('$', 'S')

    except: "error al cambiar de $ a s"

# Aplica la función a la columna 'artist'
df['artist'] = df['artist'].apply(reemplazar_simbolo_dolar)

In [49]:
df.loc[15313]


artist                                                     AsAP Rocky
url_spotify         https://open.spotify.com/artist/13ubrt8QOOCPlj...
track               F**kin' Problems (feat. Drake, 2 Chainz & Kend...
album                                 long.live.a$ap (deluxe version)
album_type                                                      album
uri                              spotify:track:4X5f3vT8MRuXF68pfjNte5
danceability                                                    0.853
energy                                                          0.693
key                                                               1.0
loudness                                                        -6.87
speechiness                                                     0.275
acousticness                                                   0.0239
instrumentalness                                                  0.0
liveness                                                         0.11
valence             

______

**4 - Filtrado de datos:**

4.1 - ¿Cuáles son los valores únicos de la columna "album_type"? Crea tres dataframes diferentes, uno para cada tipo de "album_type". ¿Cuántas canciones tenemos en cada tipo?

In [43]:
df['album_type'].value_counts()

album_type
album          14926
single          5004
compilation      788
Name: count, dtype: int64

In [60]:
valores_album = df['album_type'].unique()

In [71]:
df_album = df[df['album_type'] == valores_album[0]] # el primer valor único
df_single = df[df['album_type'] == valores_album[1]] # el segundo valor único
df_compilation = df[df['album_type'] == valores_album[2]]# el tercer valor único


In [74]:
tipo_album = [df_album, df_single,df_compilation ]

for df_tipo in tipo_album:
    print(f"En {df_tipo['album_type'].iloc[0]} hay {len(df_tipo)} canciones")

En album hay 14926 canciones
En single hay 5004 canciones
En compilation hay 788 canciones


4.2 - Usando el dataframe de los albumes que hemos creado en el ejercicio anterior. ¿Cuál es la media y la desviación estándar de "danceability", "acousticness" y "speechiness" de cada artista?. Muesrtra los resultados en tres dataframes diferentes, uno para cada métrica ("danceability", "acousticness" y "speechiness"). Ordena los resultados de mayor a menor en base a la media

In [99]:
info_de = ["danceability", "acousticness" , "speechiness"]

estadisticos_album = df_album.groupby("artist")["danceability"].agg(['mean', 'std'])

In [100]:
estadisticos_album = df_album.groupby("artist")[info_de].agg(['mean', 'std'])
estadisticos_album 

danceability           acousticness           speechiness  \
                          mean       std         mean       std        mean   
artist                                                                        
(G)I-DLE              0.796500  0.058690     0.048140  0.054108    0.128800   
*NSYNC                0.622400  0.108125     0.154031  0.180663    0.057520   
070 Shake             0.653750  0.086001     0.027217  0.027811    0.067900   
112                   0.720250  0.097025     0.129359  0.137998    0.139125   
11:11 Music Group     0.730000       NaN     0.916000       NaN    0.034000   
...                        ...       ...          ...       ...         ...   
will.i.am             0.740750  0.141959     0.013537  0.013039    0.082125   
Ángela Aguilar        0.719667  0.121970     0.462000  0.214559    0.030089   
Ñejo                  0.838000       NaN     0.177000       NaN    0.296000   
Ñengo Flow            0.719200  0.102199     0.121820  0.089099    0.232820   
Øneheart              0.303000       NaN     0.980000       NaN    0.035700   

                             
                        std  
artist                       
(G)I-DLE           0.049780  
*NSYNC             0.036145  
070 Shake          0.048658  
112                0.117117  
11:11 Music Group       NaN  
...                     ...  
will.i.am          0.050309  
Ángela Aguilar     0.003391  
Ñejo                    NaN  
Ñengo Flow         0.188013  
Øneheart                NaN  

[1989 rows x 6 columns]

In [113]:
# para todos:
info_de = ["danceability", "acousticness" , "speechiness"]
tipo_album = [df_album, df_single,df_compilation ]

for df_tipo in tipo_album:
    resultados = df_tipo.groupby("artist")[info_de].agg(['mean', 'std'])
    display(resultados)

    


danceability           acousticness           speechiness  \
                          mean       std         mean       std        mean   
artist                                                                        
(G)I-DLE              0.796500  0.058690     0.048140  0.054108    0.128800   
*NSYNC                0.622400  0.108125     0.154031  0.180663    0.057520   
070 Shake             0.653750  0.086001     0.027217  0.027811    0.067900   
112                   0.720250  0.097025     0.129359  0.137998    0.139125   
11:11 Music Group     0.730000       NaN     0.916000       NaN    0.034000   
...                        ...       ...          ...       ...         ...   
will.i.am             0.740750  0.141959     0.013537  0.013039    0.082125   
Ángela Aguilar        0.719667  0.121970     0.462000  0.214559    0.030089   
Ñejo                  0.838000       NaN     0.177000       NaN    0.296000   
Ñengo Flow            0.719200  0.102199     0.121820  0.089099    0.232820   
Øneheart              0.303000       NaN     0.980000       NaN    0.035700   

                             
                        std  
artist                       
(G)I-DLE           0.049780  
*NSYNC             0.036145  
070 Shake          0.048658  
112                0.117117  
11:11 Music Group       NaN  
...                     ...  
will.i.am          0.050309  
Ángela Aguilar     0.003391  
Ñejo                    NaN  
Ñengo Flow         0.188013  
Øneheart                NaN  

[1989 rows x 6 columns]

danceability           acousticness           speechiness  \
                          mean       std         mean       std        mean   
artist                                                                        
(G)I-DLE              0.665125  0.134657     0.025168  0.016261    0.088787   
070 Shake             0.543000  0.120542     0.081450  0.058954    0.133833   
11:11 Music Group     0.646444  0.132219     0.358353  0.370476    0.155511   
1nonly                0.773800  0.074103     0.202660  0.193339    0.215940   
2 Chainz              0.751500  0.084146     0.212150  0.211920    0.205300   
...                        ...       ...          ...       ...         ...   
will.i.am             0.777000  0.021656     0.085433  0.074003    0.102700   
Ángela Aguilar        0.631000       NaN     0.794000       NaN    0.026500   
Ñejo                  0.708778  0.071740     0.335556  0.114048    0.196800   
Ñengo Flow            0.822800  0.042056     0.109400  0.066903    0.111700   
Øneheart              0.351667  0.109253     0.917222  0.026057    0.033956   

                             
                        std  
artist                       
(G)I-DLE           0.073139  
070 Shake          0.109134  
11:11 Music Group  0.131700  
1nonly             0.093824  
2 Chainz           0.166453  
...                     ...  
will.i.am          0.067933  
Ángela Aguilar          NaN  
Ñejo               0.105816  
Ñengo Flow         0.041294  
Øneheart           0.004791  

[1355 rows x 6 columns]

danceability           acousticness            \
                             mean       std         mean       std   
artist                                                               
112                      0.786500  0.113844     0.038385  0.041741   
2 Chainz                 0.556000       NaN     0.055800       NaN   
2Pac                     0.848000  0.077904     0.038367  0.007504   
3 Doors Down             0.480000       NaN     0.018100       NaN   
A Tribe Called Quest     0.866000       NaN     0.184000       NaN   
...                           ...       ...          ...       ...   
Zendaya                  0.702000       NaN     0.005580       NaN   
Zion                     0.687000       NaN     0.209000       NaN   
back number              0.408000       NaN     0.078400       NaN   
fun.                     0.558000       NaN     0.125000       NaN   
will.i.am                0.848667  0.113147     0.191357  0.215915   

                     speechiness            
                            mean       std  
artist                                      
112                     0.288500  0.084146  
2 Chainz                0.402000       NaN  
2Pac                    0.116900  0.104324  
3 Doors Down            0.038500       NaN  
A Tribe Called Quest    0.229000       NaN  
...                          ...       ...  
Zendaya                 0.128000       NaN  
Zion                    0.101000       NaN  
back number             0.051200       NaN  
fun.                    0.032500       NaN  
will.i.am               0.061833  0.031321  

[451 rows x 6 columns]

In [115]:
    
resultados_ordenados = resultados.sort_values(('mean', info_de), ascending=False)

KeyError: ('mean', ['danceability', 'acousticness', 'speechiness'])

4.3 - Se quiere estudiar si existe una relación entre la cantidad de visitas que reciben los videos de música de un artista en una plataforma de streaming y la cantidad de likes que estos videos obtienen. Por lo tanto, querremos contestar a la siguiente pregunta: ¿Son aquellos artistas que tienen más visitas los que más likes tienen? Para solucionar este ejercicio deberéis:

- Agrupa por artista y calcular la media de visitas y de likes.
 - Quedate con los 10 artistas que más visitas han tenido.
 - Quedate con los 10 artistas que más likes han tenido.
- BONUS Haced lo mismo para los datos de tipo "single".


4.4 Ahora queremos analizar la colaboración entre artistas en la industria musical, se tienen datos de las colaboraciones realizadas en los últimos años. Se desea conocer cuáles son los artistas que han colaborado en mayor cantidad de canciones y generar un ranking con los diez primeros. ¿Qué artistas con los que más colaboraciones tienen? Para esto tendréis que:

- Filtra los datos para quedarte solo con aquellos datos donde en la columna "track" tengamos "feat".
- Cuenta cuantas canciones tiene cada artista utilizando el dataframe generado en el paso anterior.
- Debes contestar a las siguientes preguntas:
- ¿Cuáles son los diez artistas que más han colaborado según los datos que tenemos?
- ¿Cuál es el número de canciones en los que ha colaborado el artista que ocupa el primer lugar en el ranking?
- ¿Cuál es el artista que ocupa el décimo lugar en el ranking y en cuántas canciones ha colaborado?